plugins/rate_limit.go

```diff
@@ -0,0 +1,113 @@
+package plugins
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	auth "github.com/flyteorg/flyteadmin/auth"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type RateLimitExceeded error
+
+// accessRecords stores the rate limiter and the last access time
+type accessRecords struct {
+	limiter    *rate.Limiter
+	lastAccess time.Time
+}
+
+// LimiterStore stores the access records for each user
+type LimiterStore struct {
+	// accessPerUser is a synchronized map of userID to accessRecords
+	accessPerUser   map[string]*accessRecords
+	mutex           *sync.Mutex
+	requestPerSec   int
+	burstSize       int
+	cleanupInterval time.Duration
+}
+
+// Allow takes a userID and returns an error if the user has exceeded the rate limit
+func (l *LimiterStore) Allow(userID string) error {
+	l.mutex.Lock()
```

Comment:
This will cause massive contention - in effect every request is serialized for this block. I think the rate object should be accessed in parallel and then the lock should be on the lower level object.
Sync map would work right.
Otherwise this is probably a very dangerous PR

Reply:
Thanks @kumare3 and @EngHabu for the suggestion on `sync.Map`

In https://github.com/flyteorg/flyteadmin/pull/557/commits/b35114b3e55b76402fd49ba738d5599f8175e57f , a `sync.Map` is used and the lock is applied only per `UserID`. This mitigates the concern on contention, which is also the main interest of `sync.Map`

However, a "global" lock is kept at `LimitStore` level to serve the `clean` func, which periodically purge unused entries to reduce memory footprint.

Reply:
Second thoughts, actually only a single lock is needed, down to the level of `rate` object. `sync.Map` already takes care of the hard part, so all `rate` object can be accessed in parallel, even in case of `clean`.

The latest commit https://github.com/flyteorg/flyteadmin/pull/557/commits/b2383af3ac0233e2962860eb8f0c67c97d0dd880 reflected the changes. Thanks so much for spotting this out. I definitely learn a new thing today. 👍